고양이 장난감 가게 유효성 추가

BEOMKING opened the pull request:
## Summary
- 기존 장난감 가게에서 더 강한 유효성 검사를 제공합니다.
- 회원을 추가합니다.

## Detail
- 이름, 메이커, 가격을 필수 입력하도록 합니다.
- 회원은 이름, 이메일, 비밀번호를 가지고 있고 모두 필수 입력합니다.
- 조건을 어겼을 경우, 올바른 에러 메시지를 제공합니다.
- 회원은 생성, 수정, 삭제할 수 있습니다.

- 모든 코드는 TDD 로 작성합니다.
- 테스트 코드 스타일은 DCI 패턴으로 작성합니다.

BEOMKING committed:
이번주도 열심히 해보겠습니다.

Commit:
불필요한 코드 제거, 기존 Mock ControllerTest 이름 변경, 통합테스트 작성

ProductController의 통합 테스트를 만들었습니다.
생성 메서드를 실패, 빠른 성공 사이클을 수행했습니다.

Commit:
create 메서드 테스트 리팩터링

BEOMKING committed:
상품명 validation 실패하는 테스트 작성

BEOMKING committed:
상품 이름이 공백인 경우 성공하는 테스트 작성

BEOMKING committed:
상품의 이름 invalid 테스트 리팩터링

johngrib commented on app/src/main/java/com/codesoom/assignment/dto/ProductData.java:
```diff
@@ -1,26 +1,28 @@
 @Getter
 @Builder
-@NoArgsConstructor
 @AllArgsConstructor
```
이 애노테이션은 위험한 특징이 있습니다. 어떤 특징이 위험한지 한번 조사해 보세요. 그리고 lombok의 다른 애노테이션들 중에서도 또 위험한 것이 없는지도 조사해서 글로 작성해 보세요.

BEOMKING replied:
`@NoArgsConstructor` 는 기본 생성자입니다. 이 애노테이션은 `final` 사용시 컴파일 에러가 발생합니다.

`@AllArgsConstructor` 모든 필드를 가진 생성자입니다. 같은 타입의 두 필드가 있을 때, 누군가가 필드의 순서를 변경해도 코드는 에러가 발생하지 않지만 실제로는 바뀐 값들이 들어갑니다.

`@EqualsAndHashCode` 애노테이션은 `equals` 와 `hashCode` 구현을 생성해줍니다.

객체를 `Set` 자료구조에 저장하고 필드를 변경하면 `hashCode` 가 변경되어 찾을 수 없는 문제가 있습니다.

johngrib commented on app/src/test/java/com/codesoom/assignment/controllers/ProductControllerTest.java:
```diff
@@ -1,195 +1,105 @@
+    class Describe_create {
+        @Nested
+        @DisplayName("상품 정보가 주어지면")
+        class Context_with_productData {
```
컨텍스트는 테스트를 준비하는 공간입니다. "상품 정보가 주어지면"을 컨텍스트 스코프에서 준비해 보세요.

johngrib commented on app/src/test/java/com/codesoom/assignment/controllers/ProductControllerTest.java:
```diff
@@ -1,195 +1,105 @@
+            @Test
+            @DisplayName("상품을 리턴한다")
+            void It_returns_product() throws Exception {
+                create(normalInput())
+                        .andExpect(jsonPath("$.name").value(NORMAL_NAME))
+                        .andExpect(jsonPath("$.maker").value(NORMAL_MAKER))
+                        .andExpect(jsonPath("$.price").value(NORMAL_PRICE))
+                        .andExpect(jsonPath("$.imageUrl").value(NORMAL_URL))
```
`normalInput()`과 `NORMAL_NAME` ,`NORMAL_MAKER`, ... 등과의 관계가 밀접하다는 것을 이 테스트 스코프 안에서 보여줄 수 있다면 더 좋을 것 같습니다. 이들의 연관 관계는 코딩한 사람의 논리 속에서는 강하게 연결되어 있지만, 코드의 구조가 그 연관 관계를 보장하고 있지 않습니다. 처음 읽는 사람은 `normalInput` 메소드 구현을 읽어봐야 하는데, 만약 누군가가 슬쩍 `normalInput` 메소드 안쪽을 고쳐보는 일이 일어나지 않는다고 보장할 수도 없겠죠.

BEOMKING replied:
상속을 활용해 밀접하다는 것을 표현해보았습니다.

johngrib commented on app/src/test/java/com/codesoom/assignment/controllers/ProductControllerMockTest.java:
```diff
@@ -0,0 +1,195 @@
+                .price(5000)
+                .build();
+
+        given(productService.getProducts()).willReturn(List.of(product));
+
+        given(productService.getProduct(1L)).willReturn(product);
+
+        given(productService.getProduct(1000L))
+                .willThrow(new ProductNotFoundException(1000L));
+
+        given(productService.createProduct(any(ProductData.class)))
+                .willReturn(product);
+
+        given(productService.updateProduct(eq(1L), any(ProductData.class)))
+                .will(invocation -> {
+                    Long id = invocation.getArgument(0);
+                    ProductData productData = invocation.getArgument(1);
+                    return Product.builder()
+                            .id(id)
+                            .name(productData.getName())
+                            .maker(productData.getMaker())
+                            .price(productData.getPrice())
+                            .build();
+                });
+
+        given(productService.updateProduct(eq(1000L), any(ProductData.class)))
+                .willThrow(new ProductNotFoundException(1000L));
+
+        given(productService.deleteProduct(1000L))
+                .willThrow(new ProductNotFoundException(1000L));
```
given 을 통한 전제가 필요 이상으로 많아 테스트를 통해 버그를 발견하기 힘든 상황입니다. 지금 상태에서는 `productService` 구현에 치명적인 문제가 발생한다 해도 이 테스트는 고장난 화재경보기처럼 언제나 통과하기만 할 거에요.

BEOMKING replied:
맞습니다. 혹시 이런 방법은 어떨까요?

1. 가장 외부의 Controller 를 통합 테스트로 작성한다. Controller 에서 하고자 하는 기능을 작성해서 실패 -> 빠른 성공 과정을 진행한다.
2. 이후 리팩터링 과정으로 Service 와 Repository 를 Mock 단위 테스트를 해서 기능을 완성한다.

Service 와 Repository 에서 고장난 화재경보기가 되어도 결국 Controller 에서 테스트를 완성하기 위해 고장난 화재경보기를 고칠 수 있지 않을까 생각합니다.

johngrib replied:
말씀하시는 방법도 많이들 사용하죠. 이번주는 그렇게 진행해 보는 것도 좋겠습니다.

BEOMKING replied:
오! 많은 분들이 사용하는 방법이었군요!
뭔가 자신감이 생깁니다 ㅎ

johngrib replied:
많이 사용하긴 하는데 좋은 선택이 아닌 경우도 있습니다. 그래서 잘 생각해야 하죠.
저는 리스크가 있는 프로젝트일수록 mock을 쓰는 건 가능한 한 최소한도로 하는 것이 바람직하다고 생각합니다.
지금은 컨트롤러가 유일한 서비스의 클라이언트이기 때문에 이 방법이 통하는 거죠.

서비스가 컨트롤러 없이/다양한 종류의 컨트롤러에 대해 스스로 제대로 동작한다는 걸 보여주는 것이 필요한 단계가 되면 mock 없애느라 굉장히 고생스럽게 됩니다.
mock이 있어도 보장해 주는 것이 껍데기에 가까운 코드들에서 멈추는 경우가 있기 때문이죠.

BEOMKING replied:
mock의 사용을 고민할 때, 어떤 부분을 고려해야 하는지 감이 오는 것 같습니다! 감사합니다 ㅎㅎ

johngrib replied:
좋은 주제로 고민을 하고 계신 것 같아 기쁩니다. 다른 분의 PR에서 나눈 이 대화를 @BEOMKING 님이 읽어보시는 것도 사고를 더 넓게 하시는 데 도움이 될 것 같습니다.

https://github.com/CodeSoom/spring-week3-assignment-1/pull/58#discussion_r845062254

johngrib reviewed (commented):
오랜만입니다 :-)

BEOMKING committed:
상속을 활용해 Normal input과 결과값의 관계를 표현

Commit:
이름이 빈 값을 검증하는 기능

Commit:
이름 범위 Validation

Commit:
이름 Validation 테스트 리팩터링

Commit:
메이커 Validation 생성 및 테스트

Commit:
maker validation 잘못된 애노테이션 수정, 가격 validation 추가

Commit:
가격 Validation 추가

가격이 null 인 경우, 음수인 경우, 한계치를 초과한 경우를 테스트
추가했습니다.

Commit:
빈값 테스트 리팩터링, 이름 길이 테스트 추가

Commit:
메이커, 가격 Validation 추가

메이커의 길이, 가격이 비어있을 때, 음수 일때, 한계치를 넘어갔을 때 검증

Commit:
EOF 추가

BEOMKING commented on app/src/test/java/com/codesoom/assignment/controllers/ProductControllerTest.java:
```diff
@@ -135,5 +135,154 @@ void It_returns_exceptionResponse(String input) throws Exception {
+            }
+        }
+
+        @Nested
+        @DisplayName("상품 가격이 한도를 초과하면")
+        class Context_with_priceExceedLimit extends Normal {
```
외부에서 들어오는 요청에 대한 Validation(DTO) 을 모두 처리하다 보니 각 Context 에 비슷한 코드들이 계속 작성되고 필드의 양이 많지 않음에도 테스트 코드량이 많이 늘어나는군요.

johngrib replied:
테스트 코드를 작성할 때에는 프로덕션 코드와 다른 관점에서 접근할 필요도 있습니다. 예를 들어 프로덕션 코드에서 핵심적인 원칙 중의 하나인 DRY는 테스트코드에서는 상대적으로 힘을 잃습니다. 테스트코드는 지정된 범위 안쪽에서 해당 테스트를 논리적이면서 가독성있게 작성하는 것이 엄청나게 중요하기 때문이에요. 그래서 중복 제거가 테스트코드에서는 최우선이 아닐 수 있습니다. 중복이 좀 되더라도 테스트코드의 논리를 읽는 사람에게 명확하게 전달하면서 이 파일 저 파일 열어보지 않고도 화면 하나 안에서 이해시키는 것이 중요한 거죠.

테스트 코드 하나는 스스로 설명적일수록 더 높은 완성도를 갖게 될 개연성을 얻습니다.
시간을 내어 이 말을 곱씹어보세요.

---

그 다음으로 중요한 것은 테스트코드와 테스트코드가 주어졌을 때 어떤 것이 더 중요한지를 판단하는 것입니다. 시간은 한정되어 있고, 우리는 모든 것을 테스트할 수 없습니다. 경우에 따라 어떤 테스트 코드는 트레이드 오프를 고려해 작성하지 않기로 선택할 수도 있는 것입니다.

제 기준에서 가장 중요한 테스트 코드는 "과거의 버그가 되살아났을 때" 그것을 감지해주는 테스트 코드입니다. 그 외에는 중요도가 떨어지죠.

이 테스트 코드의 중요성에 대해 열심히 생각해보세요. 어쩌면 별로 중요한 테스트 코드가 아닐 수도 있습니다.

BEOMKING committed:
TDD를 다시 하기 위해 코드 제거

서비스, 테스트 코드 제거, 컨트롤러 로직 제거

BEOMKING committed:
오타 수정

johngrib commented on app/src/test/java/com/codesoom/assignment/dto/ProductDataTest.java:
```diff
@@ -0,0 +1,116 @@
+        assertEquals(1, constraintViolations.size());
+        assertEquals("이름의 길이가 범위를 벗어납니다.", constraintViolations.iterator().next().getMessage());
+
+        productData = new ProductData("길이가 범위를 넘는 값을 가진 이름임ㅇ", NORMAL_MAKER, NORMAL_PRICE, NORMAL_URL);
```
오타가 있군요.

johngrib commented on app/src/test/java/com/codesoom/assignment/dto/ProductDataTest.java:
```diff
@@ -0,0 +1,116 @@
+        constraintViolations = validator.validate(productData);
+
+        assertEquals(1, constraintViolations.size());
+        assertEquals("이름은 필수값입니다.", constraintViolations.iterator().next().getMessage());
```
메시지를 비교하는 것보다 더 나은 방법이 있을 것 같군요.

BEOMKING replied:
13713d7
메시지를 비교하는 방법에서 조건을 위반했는지(validate 결과로 `ConstranintViolation` 이 만들어지는지)를 테스트하는 방법으로 변경했습니다.

메시지를 비교하는 방법은 프로덕션 코드와 커플링이 생겨서 메시지를 변경하고자 하면 테스트 코드를 바꿔야 하는 상황이 올 듯합니다. (Controller 통합 테스트에도 적용되는 내용이겠네요!)
Validation이 조건을 위반하는지만 테스트해도 충분할 것 같습니다.

종립님은 어떻게 생각하시나요?

johngrib replied:
이 정도로도 충분해 보입니다. 뭔가 더 추가하는 건 나중에 버그가 이상동작이 발견된 다음에도 괜찮아 보입니다.

johngrib commented on app/src/test/java/com/codesoom/assignment/dto/ProductDataTest.java:
```diff
@@ -0,0 +1,116 @@
+    }
+
+    @ParameterizedTest
+    @ValueSource(strings = {"", " "})
```
탭 문자, 개행 문자, 2개 이상의 스페이스 등도 추가할 수 있겠습니다.

BEOMKING replied:
생각보다 고려할 것이 많았군요!

BEOMKING committed:
create 테스트 실패하는 테스트 코드 작성

BEOMKING committed:
오타 수정

BEOMKING committed:
공백 문자 테스트 조건 추가

BEOMKING committed:
상품 서비스 생성 로직 구현

상품 정보를 상품으로 바꾸는 메서드 구현
컨트롤러와 생성자에 의존성 주입
DTO, Service 테스트 작성

BEOMKING committed:
상품 Repository 저장 메서드 테스트

Commit:
상품 조회 기능 구현

상품 컨트롤러, 서비스, 리포지토리 구현 및 테스트

Commit:
상세 조회 테스트 구현 및 한글 문제로 인해 -> 영어로 인풋 변경

Commit:
Service 테스트 안 올린 것 추가

Commit:
식별자를 가진 상품이 없을 경우 실패하는 테스트

Commit:
없는 상품을 조회하는 경우 구현

Commit:
상품 목록 조회 Controller 실패 -> 성공 과정 수행

Commit:
상품 전체 조회 Service, Repository 구현

Commit:
DTO Validation 테스트, 오타 수정

메시지와 정확히 일치하는 지 테스트하는 방법에서 예외 메시지가 있는 지
확인하는 테스트로 수정

Commit:
DB 초기화용 deleteAll 구현, BeforeEach에 영향 받는 테스트 리팩터링

johngrib commented on app/src/main/java/com/codesoom/assignment/application/ProductService.java:
```diff
@@ -1,64 +1,14 @@
-@Transactional
-public class ProductService {
-    private final ProductRepository productRepository;
+public interface ProductService {
```
좋습니다. 한 단계 더 나아가서 이 `ProductService`를 둘로 나누어 보는 연습을 해보는 것도 좋겠습니다.

조회를 담당하는 `ProductQueryService`와 값의 입력과 업데이트, 삭제 등을 담당하는 `ProductCommandService`로 쪼개보는 거죠. 조회를 따로 분리해두면 복잡도가 줄어 한층 관리하기 편해질 겁니다.

물론 지금은 메소드가 몇 개 안 되니까 굳이 나눌 필요는 없을 수 있습니다. 하지만 연습이니까요 :-)

BEOMKING replied:
평소에 프로젝트를 할 때에도 조회 기능이 많았던 기억이 있습니다.

지금 메소드가 많지 않음에도 분리하니까 테스트 코드의 양이 줄어들고 관리하기 좋아진 듯 합니다.

johngrib replied:
좋다고 느끼시다니 기쁩니다.
저도 평소에 회사에서 일을 할 때 조회 기능은 따로 모아두는 편입니다. 그러면 해당 파일을 열었을 때 "아 여기에는 조회만 있으니까 데이터를 변경하는 코드는 없어" 라고 안심하며 코드를 읽을 수 있죠. 전 이것이 꽤 좋은 습관이라고 생각합니다.

johngrib commented on app/src/main/java/com/codesoom/assignment/application/ToyService.java:
```diff
@@ -0,0 +1,36 @@
+        return productRepository.save(productData.toProduct());
+    }
+
+    @Override
+    public Product findById(Long id) {
+        return productRepository.findById(id)
+                .orElseThrow(() -> new ProductNotFoundException(id));
+    }
```
이렇게 예외를 던지는 경우에는 꼭 JavaDoc을 추가해서 `@throws`의 내용을 작성해 줍시다. 항상 그럴 수 있도록 습관으로 만들어 두세요.

BEOMKING replied:
인터페이스에 JavaDoc 을 추가했습니다!

johngrib commented on app/src/main/java/com/codesoom/assignment/application/ToyService.java:
```diff
@@ -0,0 +1,36 @@
+                .orElseThrow(() -> new ProductNotFoundException(id));
+    }
+
+    @Override
+    public Collection<Product> findAll() {
+        return productRepository.findAll();
+    }
```
단순한 `findAll` 호출이라 좀 싱겁군요. 여유가 있다면 다음 두 가지를 추가 과제라고 생각하고 시도해 보세요.

1. Product에 status를 추가하고, 다양한 상태를 지정할 수 있게 해봅니다. (판매중, 품절됨, 관리자에 의해 숨겨짐, ... 등의 다양한 상태가 있을 수 있겠죠.)
2. `findAll`을 호출할 때 숨겨진 상태이거나 품절된 상태의 상품은 제외하고 응답하도록 수정해 보세요. 이렇게 하면서 메소드 이름을 변경하거나, 새로운 메소드를 추가하는 방식으로 작업할 수도 있겠죠.

BEOMKING replied:
상품의 상태를 나타내는 Status 를 만들어서 추가 과제를 진행하고 있습니다!

Status가 빈 값을 허용하지 않는 조건을 주고 추가하면서 필드 검증을 하는 많은 테스트 코드들에 영향이 가네요 ㅠ

- Controller 테스트 준비에 Status 가 없어 모든 테스트가 깨지거나
- Controller, ProductData 에서 Validate 할 때
- Repositroy 에 저장할 때 
 
등등
 
필드가 추가되었으니 테스트가 통과되지 않고 수정해야 할 것이 많은 게 당연한 건지, 테스트 코드를 리팩터링해야 할 신호인건지 잘 모르겠습니다.

이런 경우 관리가 힘들어진다고 생각하는데 이럴 때 사용할만한 좋은 방법이 있을까요?

johngrib replied:
테스트 코드에 영향이 가서 번거로우셨겠습니다만... 사실 그것을 노렸습니다! (죄송합니다!)
이런 경험을 얻게 되면 어떤 방향으로 테스트코드를 만들어야 나중에 기능이 확장되면서도 테스트 코드가 와장창 깨지지 않을까를 고민하게 되죠. 그래서 코딩을 하면서도 (아직 없는) 테스트코드를 꾸준히 의식할 필요를 느끼게 됩니다. 물론 테스트코드를 먼저 작성하는 TDD를 할 때에도 마찬가지입니다. 먼저 테스트 코드를 만들긴 했지만 이 테스트 코드는 변경에 취약한가? 아니면 변경에도 유연하게 대처할 수 있는 테스트 코드인가? 를 생각하며 작업하게 되는 거죠.

관리가 힘들어지는 경우에 제가 쓰는 방법이 하나 있습니다. 물론 절대적인 건 아니니 쓸모에 따라 제 방법을 적당히 자기 것으로 해 보세요.

제 방법은 삭제하기 쉽게 만드는 겁니다. 우리가 하고 있는 D-C-I 스타일의 테스트 코드가 이 방법을 적용하기에 딱 좋습니다. 계층구조로 테스트 코드를 작성해 놓고 테스트가 깨질만한 문제가 생겼을 때 특정 스코프를 범위로 잡아서 (마치 어떤 디렉토리를 삭제하면 자식 디렉토리들도 한꺼번에 삭제되는 것처럼) 한꺼번에 삭제할 수 있게 하는 겁니다.

여기에서 핵심은 삭제가 아닙니다. "삭제 가능성"이죠. 경험상 삭제 가능성이 뛰어나면 고치기도 편했습니다. 삭제같은 극단적인 방법이 기준이 되어서 그런 것 같기도 합니다.

테스트 코드는 리팩터링하기보다는 코드를 검증하는 날카로운 논리가 잘 벼려져 있는지를 확인하는 것이 좋습니다. 위의 코멘트에서도 이야기했지만 테스트 코드에 대해서는 코드의 효율이나 중복 제거 등을 최우선 목표로 하지 않도록 해보세요. 그런 것들은 테스트 코드의 제 1 중요도가 아닙니다. 코드를 수정했을 때 테스트 코드를 이렇게 저렇게 고치는 것보다 더 중요한 것은 테스트 코드의 "논리"가 수정한 코드의 "논리"를 엄격하게 잘 검사하고 있는지입니다. 논리가 깨졌다면, 혹은 전제가 잘못됐다면 그 테스트코드는 고쳐서 다시 돌아가게 만들었더라도 나중에 읽는 사람의 시간만 버리는 쓸데없는 테스트 코드가 되는거죠. 논리를 검사하고 논리를 드러내는 방식으로 다시 만들거나 고쳐야겠다고 생각하며 접근하는 방법을 고려해 보세요.

이걸 생각하고 테스트코드를 관리하도록 해보세요. 그렇다면 새로운 것을 깨닫게 되실 수도 있을 겁니다.

BEOMKING replied:
이번 주 과제를 진행하면서 `이제 좀 작성할 줄 알겠다` 라고 생각했었는데 추가 과제를 진행하니까 아직 멀었다는 것을 알았습니다. 기능 추가를 해보기 정말 잘한 것 같습니다. 감사합니다!! ㅎ

johngrib commented on app/src/test/java/com/codesoom/assignment/AppTest.java:
```diff
@@ -9,7 +9,6 @@ class AppTest {
         App classUnderTest = new App();
-        assertNotNull(classUnderTest.getGreeting(), "app should have a greeting");
-        assertEquals(classUnderTest.getGreeting(), "Hello, world!");
+        assertEquals(classUnderTest.getGreeting(), "안녕하세요! 잘 부탁드립니다.");
```
이 테스트는 이 메소드를 왜 테스트하는지를 구체적으로 생각하는 것이 아주 중요합니다. 그러다보면 뭔가 깨달을 수 있을 거에요. 그리고 테스트 검증 방식을 다르게 고치고 싶어질 겁니다.

BEOMKING replied:
이 테스트는 `@SpringBootTest`를 붙혀서 스프링 애플리케이션이 정상적으로 작동하는 지 테스트하기 위해 사용하는 줄 알고 있었습니다. getGreeting 을 왜 테스트하는 지는 잘 모르겠습니다.

johngrib replied:
이미 알고 계십니다! 이 메소드는 `스프링 애플리케이션이 정상적으로 작동하는 지 테스트하기 위해 사용하는` 것이죠!

그래서 이 메소드가 "안녕하세요"라는 내용의 문자열을 리턴하는지 "안녕히 계세요"라는 내용의 문자열을 리턴하는지는 전혀 중요하지 않습니다. 문자열이 리턴된다는 게 중요한거고, 그걸 통해 우리가 스프링 애플리케이션이 정상적으로 작동한다는 걸 깨닫게 되는 것이 중요한 거죠. 말하자면 이 테스트의 목적은 "이 애플리케이션이 멀쩡히 돌아간다는 증거로 인사 문자열을 리턴하는 메소드를 호출해 보는 것"이라 할 수 있습니다.

이제 `@DisplayName`에 무엇을 적고 `assert`에서 무엇을 검사해야 할 지 생각해볼 때가 됐습니다. `assert`에서는 에러가 안 나는지, null이 아닌지만 검사해도 충분하겠죠.

Commit:
ProductData Validation 테스트 커플링 제거

메시지를 테스트하는 방법 -> 예외 발생 체크 테스트

Commit:
deleteAll 테스트가 제대로 검증하지 못하던 에러 수정

준비된 상황을 실행하지 않아도 테스트가 통과하는 문제가 있어 준비한
상황이 잘 실행되었는지 검증하는 코드 추가

Commit:
ProductService의 조회 메서드 분리

Commit:
ProductCommandService 분리

상품의 변형을 주는 기능들을 가진 인터페이스를 분리했습니다.
조회 기능이 분리됨으로써 컨트롤러의 조회 기능의 의존을
ProductQueryService로 변경했습니다.

Commit:
Transactional 추가

상품이 변경되었을 때, 문제가 발생했을 경우 롤백 처리하기 위해
추가했습니다.

Commit:
상품의 상태 추가

Enum Status 생성
상품에 status 필드 생성

Commit:
ProductQueryService 판매중인 상품 목록을 조회하도록 변경

findAll 주석 변경(판매중인)
findAll 판매중인지 확인해서 리턴하도록 변경
Product와 Status에 판매중인지 확인하는 메서드 추가
Service findAll 테스트

Commit:
상품의 필드 추가에 영향을 받는 테스트 코드 수정

상품 정보, 상품에 status 필드가 빈 값이 될 수 없도록 설정됨에 따라
생성자를 사용하는 테스트 코드들에 status 필드 추가

BEOMKING commented on app/src/test/java/com/codesoom/assignment/dto/ProductDataTest.java:
```diff
@@ -101,18 +103,33 @@ void priceCannotBeNegativeOrExceedTypeLimit() {
     void productDataCanConvertProduct() {
-        assertEquals(new Product(NORMAL_NAME, NORMAL_MAKER, NORMAL_PRICE, NORMAL_URL),
-                new ProductData(NORMAL_NAME, NORMAL_MAKER, NORMAL_PRICE, NORMAL_URL).toProduct());
+        assertEquals(new Product(null, NORMAL_NAME, NORMAL_MAKER, NORMAL_PRICE, NORMAL_URL, Status.SALE),
+                new ProductData(NORMAL_NAME, NORMAL_MAKER, NORMAL_PRICE, NORMAL_URL, SALE).toProduct());
```
상태 필드가 빈 값을 허용하지 않도록 추가되면서 필드 검증을 하는 많은 테스트 코드들에 영향이 가네요 ㅠ

이런 경우 관리가 힘들어진다고 생각하는데 이럴 때 사용할만한 좋은 방법이 있을까요?

현재는 상품과 관련된 테스트 데이터를 저장하는 클래스를 만드는 방법 밖에 생각이 나지 않는군요.

Commit:
상품의 조회 결과로 boolean status 값을 리턴하던 에러 수정

Commit:
불필요한 코드 제거, 설명 통일

Commit:
status 검증 통합, 단위 테스트 추가

Commit:
품절 상품 목록 조회 기능 실패하는 테스트

BEOMKING committed:
품절된 상품 목록 조회 기능 구현

BEOMKING committed:
QueryServiceTest 리팩터링

johngrib commented on app/src/main/java/com/codesoom/assignment/application/ProductCommandService.java:
```diff
@@ -0,0 +1,13 @@
+ * 상품에 변형을 주는 기능을 제공합니다.
+ */
+public interface ProductCommandService {
+    Product create(ProductData productData);
```
update, delete 도 추가할 수 있을 것 같습니다. 아, 아직은 과제에 포함되어 있지 않은가요?

BEOMKING replied:
넵 곧 진행할 예정입니다!

johngrib commented on app/src/main/java/com/codesoom/assignment/application/ProductQueryService.java:
```diff
@@ -0,0 +1,34 @@
+ * 상품의 조회 기능을 제공합니다.
+ */
+public interface ProductQueryService {
+    /**
+     * 식별자를 가진 상품을 리턴한다.
+     *
+     * @param id 식별자
+     * @return 상품
+     * @throws ProductNotFoundException 식별자를 가진 상품이 없으면 예외를 던진다
+     */
+    Product findById(Long id);
```
이렇게 한번 해보세요.
```suggestion
    /**
     * 상품을 찾아 리턴한다.
     *
     * @param id 식별자
     * @return 상품
     * @throws ProductNotFoundException 상품을 찾지 못한 경우
     */
    Product findById(Long id);
```

한편 `@throws`에 있는 `상품이 없으면`에 대해 곰곰히 생각해 보세요. 컴퓨터 시스템은 다양한 이유로 예외가 던져질 수 있습니다. 어떤 경우에는 DB에 상품이 있어도 정전이 되거나 네트워크 딜레이가 길어지거나 해서 있는 상품이 없다고 예외가 던져지기도 할 겁니다. 이런 경우를 잘 생각해보면 OS나 유명한 애플리케이션에서는 "찾지 못했습니다"라고 변명하는 상황을 본 적이 있을 겁니다. 절대로 "없다"라고 하지 않습니다. 이런 경우를 "겸손한 애플리케이션"이라고 부릅니다.

BEOMKING replied:
큰 차이가 있네요!

Commit:
AppTest에서 애플리케이션이 정상 작동을 확인하는 테스트 작성

BEOMKING committed:
Update app/src/main/java/com/codesoom/assignment/application/ProductQueryService.java

주석 변경

Co-authored-by: JohnGrib <johngrib82@gmail.com>

Commit:
상품 삭제 기능 실패 -> 빠른 성공 테스트 작성

Commit:
찾는 상품을 삭제기능 서비스, 리포지토리 구현 후 컨트롤러 성공 완성

Commit:
Merge branch 'BEOMKING' of https://github.com/BJP-Learning-Space/spring-week5-assignment-1 into BEOMKING

Commit:
상품 목록 제거 실패하는 테스트

Commit:
삭제할 상품 목록을 받아 삭제하는 기능 구현

BEOMKING committed:
컨트롤러 상품 정보 변경 기능 실패 -> 빠른 성공 테스트

BEOMKING committed:
상품 정보 변경 기능 구현

BEOMKING committed:
유저와 상품을 분리하기 위해 패키지 구조 변경

BEOMKING committed:
컨트롤러 유저 생성 기능 실패 -> 성공 테스트

johngrib commented on app/src/test/java/com/codesoom/assignment/AppTest.java:
```diff
@@ -1,14 +1,18 @@
-    void appHasAGreeting() {
-        App classUnderTest = new App();
-        assertEquals(classUnderTest.getGreeting(), "안녕하세요! 잘 부탁드립니다.");
+    @DisplayName("애플리케이션은 정상적으로 작동한다")
+    void appWorksFine() {
+        App app = new App();
+
+       assertNotNull(app.getGreeting());
```
좋습니다. 다만 `@SpringBootTest`를 활용한다면 좋겠죠. `new`가 눈에 걸리는군요.

BEOMKING replied:
`@SpringBootTest`를 쓰고 활용하지 않았군요,,

johngrib commented on app/src/main/java/com/codesoom/assignment/controllers/ProductController.java:
```diff
@@ -51,4 +52,10 @@ public Collection<Product> getAll() {
+    @ResponseStatus(HttpStatus.NO_CONTENT)
+    public void delete(@PathVariable("id") Long id) {
+
+    }
```
이런 경우에는 다음과 같이 TODO 주석을 추가해 보세요. 나중에 TODO만 모아서 보기 좋을 겁니다.

```
// TODO: delete 메소드 구현
```

만약 이 메소드가 구현되지 않은 상태로 배포되는 것이 곤란하다면 이 메소드가 호출될 때 무조건 예외를 던지도록 코딩해두는 것도 방법입니다. 지금은 프로젝트가 작고 파일도 몇 개 안 되니 이런 문제는 쉽게 발견할 수 있지만, 만약 파일이 몇천개가 된다면 이야기가 다를 수 있겠죠. 그런 경우엔 예외를 통해 찾아내는 방법이 가장 빠른 방법 중 하나이죠.

BEOMKING replied:
업데이트 기능을 구현하다 `실패하는 테스트 -> 빠른 성공 테스트` 를 작성한 후 제 기능을 동작하도록 구현했는데 그걸 컨트롤러에 적용하는 것을 깜빡한 일이 생겼습니다. 일단 테스트를 통과되도록 만든 후라 문제를 체크하지 못했네요.

이런 경우 말씀해주신 예외를 던지도록 작성해두는 것이 큰 도움이 되겠네요! 주석 또한 활용해보겠습니다.

johngrib commented on app/src/main/java/com/codesoom/assignment/application/ToyCommandService.java:
```diff
@@ -20,4 +20,9 @@ public ToyCommandService(ProductRepository productRepository) {
+    @Override
+    public void deleteById(Long id) {
+        productRepository.deleteById(id);
+    }
```
이 메소드에서 생각해볼만한 문제가 하나 있습니다. 만약 "삭제를 수행했을 경우 몇 건이 성공했는지를 응답으로 받고 싶다"는 요구사항이 추가된다고 상상해 보세요. 어떻게 대응할 수 있을까요?

BEOMKING replied:
우선 현재 삭제 기능을 제공하는 메소드에 대한 테스트를 수정하고 구현하는 방식으로 진행해보았습니다.

그런데 요구사항이 자주 바뀌는 상황이라면, 바뀔 때마다 매번 삭제 기능과 관련된 모든 테스트와 구현을 고칠 수는 없지 않을까? 하는 생각이 들었습니다.

johngrib replied:
좀 뜬금없어 보일 순 있지만 이번 기회에 `deleteById`의 안쪽에서 어떻게 돌아가고 있는지를 조사해볼 기회이기도 합니다.

BEOMKING replied:
deleteById는 내부적으로 null을 체크하는 로직과 findById를 사용하고 있었네요!

```
Assert.notNull(id, ID_MUST_NOT_BE_NULL);

delete(findById(id).orElseThrow(() -> new EmptyResultDataAccessException(
			String.format("No %s entity with id %s exists!", entityInformation.getJavaType(), id), 1)));
```

BEOMKING committed:
스프링부트 정상 동작 테스트 리팩터링

BEOMKING committed:
빠른 성공 -> 리팩터링을 구현하고 적용하지 않은 코드 적용

BEOMKING committed:
삭제시 성공한 개수를 리턴하도록 테스트 수정

BEOMKING committed:
커맨드 서비스의 삭제 메서드 시그니처 변경, 구현 변경

BEOMKING committed:
컨트롤러 상품 삭제 개수 구현

BEOMKING committed:
여러개 삭제 테스트 수정

BEOMKING committed:
컨트롤러 서비스 여러개 삭제하고 개수를 리턴하도록 테스트 수정 및 구현

johngrib commented on app/src/test/java/com/codesoom/assignment/product/application/ProductCommandServiceTest.java:
```diff
@@ -121,16 +121,34 @@ class Describe_delete {
 
-                verify(productRepository).deleteById(BASIC_ID);
+            @Test
+            @DisplayName("0을 리턴한다")
```
좋습니다. 아주 명쾌하고 오해의 소지가 전혀 없습니다.

johngrib commented on app/src/main/java/com/codesoom/assignment/product/application/ToyCommandService.java:
```diff
@@ -30,8 +32,15 @@ public Product update(Long id, ProductData data) {
     @Override
-    public void deleteById(Long id) {
-        productRepository.deleteById(id);
+    public int deleteById(Long id) {
+        Optional<Product> storedProduct = productRepository.findById(id);
+
+        if (storedProduct.isPresent()) {
+            productRepository.delete(storedProduct.get());
+            return 1;
+        }
+
+        return 0;
```
만약 이 메소드의 리턴값이 음수일 수 없다는 것을 아주 강력하게 표현하고 싶다면 어떤 방법을 쓸 수 있을까요?

BEOMKING replied:
주석으로 표현하기에는 약한 것 같고 컴파일이 되지 않도록 하면 좋을 것 같은데..음

johngrib replied:
여러가지 방법이 있겠지만 리턴 전에 체크해서 예외를 던지는 방법도 생각해 볼 수 있겠죠.

johngrib commented on app/src/test/java/com/codesoom/assignment/product/controllers/ProductControllerTest.java:
```diff
@@ -517,8 +518,8 @@ void It_remove_products() throws Exception {
-                        .andExpect(status().isNoContent())
-                        .andExpect(jsonPath("$").doesNotExist());
+                        .andExpect(status().isOk())
+                        .andExpect(jsonPath("$").value(3));
```
이렇게 구현이 안 된 경우에는 "일단 테스트를 통과하게" 하는 게 아니라 "무조건 테스트를 실패하게" 하는 것이 더 바람직합니다.

BEOMKING replied:
실패 -> 성공 -> 리팩터링 과정을 너무 지키려했었네요. 구현이 안된 경우에는 테스트가 실패하도록 작성하겠습니다.

johngrib reviewed (commented):
이번주 과제를 훌륭히 잘 해내고 계신 것 같습니다. 그러고보니 벌써 토요일 밤이로군요. 주간 회고를 작성해 공유해 주세요.

BEOMKING committed:
deleteById 내부 구현 확인 테스트 작성

id에 맞는 상품이 없으면 어떤 예외를 발생시키는지
id가 null인 경우엔 어떤 예외를 발생시키는지 확인하는 테스트를 작성했습니다.

johngrib reviewed (commented):
이번주 고생 많으셨습니다. 추가 과제까지 열의를 보이며 열심히 해주셔서 리뷰하면서도 재미있었네요. 주간 회고를 작성해 공유해 주세요. 다음에 또 만납시다.
